rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+dynamism, almost all ops can unambiguously infer their output types from their
+inputs; with dynamism, this is not the case) and complicating the implementation
+(checking constraints become more involved).
+
+Much of this is already implemented in the StableHLO dialect (e.g. verifiers and
+shape functions already support dynamism in a fairly robust manner, although
+some open questions still remain), but formalizing this area is left for future
+work, because that would require formalizing the general notion of shape
+inference which represents a significant amount of work.
+
+<a id="o5" /> **(O5)** [Value inference](https://github.com/tensorflow/tensorflow/blob/cc71ba69fa9d25a21009b6e377f3dc3d1323aa6c/tensorflow/compiler/xla/client/value_inference.h)
+is an algorithm implemented in the XLA compiler. It "analyzes values in XlaOp
+answers following questions: 1) What's the upper-bound of each value in a
+tensor, 2) What's the lower-bound of each value in a tensor, 3) What's the
+constant value of each tensor, 4) Whether or not each value in a tensor
+is dynamic". This algorithm is used in the TF/XLA bridge to automatically
+compute bounds for bounded types in HLO programs.
+
+Similarly to shape inference, this area is out of scope for this RFC.
+From the implementation standpoint, value inference is currently not available
+for the StableHLO dialect, but this may change in the future depending on
+community feedback.
+
+<a id="o6" /> **(O6)** Shape specialization is a practically important problem
+because not all consumers support dynamic shapes. In the StableHLO repository,
+there are `--stablehlo-refine-shapes` and `--stablehlo-canonicalize-dynamism`
+passes which address this problem. Furthermore, based on a semi-formal proof and
+the experience with JAX native serialization, the author believes that these
+passes are guaranteed to fully specialize dynamic StableHLO programs which only
+involve shape polymorphism and have static arguments to static StableHLO programs.
+
+However, similarly to formalizing shape inference, formalizing shape
+specialization is a significant amount of work which is not on the critical
+path to [StableHLO v1.0](../docs/roadmap.md), so it is left for future work.
+
+<!-- markdownlint-disable line-length -->
+## <a id="p1" /> (P1) Move TensorFlow-specific operations out of StableHLO
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** The MHLO and CHLO dialects have been initially co-designed with
+the TensorFlow dialect, the MLIR-based TF/XLA bridge and TensorFlow's
+[KernelGen toolchain](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/mlir/tools/kernel_gen).
+This has been inherited by the StableHLO repository when the StableHLO dialect
+was bootstrapped from MHLO and the CHLO dialect was moved to the StableHLO
+repository.
+
+As a result, there are 2 StableHLO ops (`compute_reshape_shape` and
+`cstr_reshapable`) as well as 4 CHLO ops (`dynamic_reshape`,
```

Comment:
It seems that JAX also uses `dynamic_reshape`.

Reply:
JAX is using the `stablehlo.dynamic_reshape`, but this is talking about `chlo.dynamic_reshape`. The naming of this op is really confusing, all things considered.